src/main/java/com/ncc/aif/ValidateAIF.java

```diff
@@ -574,4 +592,183 @@ public Resource validateKBAndReturnReport(Model dataToBeValidated, Model union)
+            }
+        }
+
+        // Dump stats to the specified PrintStream
+        private void dumpStats(PrintStream out) {
+            final SortedMap<Integer, ExecStatistics> savedStats = new TreeMap<>();
+            final SortedSet<Map.Entry<Integer, ExecStatistics>> sortedStats = new TreeSet<>(
+                    (e1, e2) -> Long.compare(e2.getValue().getDuration(), e1.getValue().getDuration()));
+            List<ExecStatistics> stats = ExecStatisticsManager.get().getStatistics();
+            stats.forEach(n -> {
+                if (n.getDuration() > durationThreshold) {
+                    savedStats.put(stats.indexOf(n), n);
+                }
+            });
+            if (savedStats.isEmpty()) {
+                out.println("There were no queries that took longer than " + durationThreshold + "ms (of "
+                        + stats.size() + " queries overall).");
+            }
+            else {
+                out.println("Displaying " + savedStats.size() + " slow queries (of "
+                        + stats.size() + " queries overall).");
+                sortedStats.addAll(savedStats.entrySet());
+                sortedStats.forEach(n -> dumpStat(n.getKey(), n.getValue(), out));
+            }
+        }
+
+        // Dump a single query's statistics to the specified PrintStream
+        private void dumpStat(Integer queryNum, ExecStatistics queryStats, PrintStream out) {
+            out.println("\n\nQuery #" + queryNum+1);
+            out.println("Label: " + queryStats.getLabel());
+            out.println("Duration: " + queryStats.getDuration() + "ms");
+            out.println("StartTime: " + new Date(queryStats.getStartTime()));
+            out.println("Context node: " + queryStats.getContext().toString());
+            out.println("Query Text: " + queryStats.getQueryText().replaceAll("PREFIX.+\n", ""));
+        }
+    }
+
+    /**
+     * A statistics collector for use in profiling TopBraid-based SHACL validation.  Typical usage is to call
+     * {@link #startCollection()} and {@link #endCollection()} to bound statistics collection,
+     * then call {@link #dump(String)} to dump slow query statistics to <filename>-stats.txt.
+     *
+     * This statistics collector progressively dumps slow queries to stdout, which is useful if you suspect
+     * the validation will not complete due to out of memory or other error conditions.
+     */
+    private static class ProgressiveStatsCollector implements ExecStatisticsListener {
+
+        private final int durationThreshold;
+        private final SortedMap<Integer, ExecStatistics> savedStats = new TreeMap<>();
+
+        /**
+         * Creates a statistics collector that progressively dumps queries slower than [threshold] ms to stdout.
+         * @param threshold the threshold definition of a slow query for this statistics collector
+         */
+        ProgressiveStatsCollector(int threshold) {
```

Comment:
There is a lot of duplication of code in ProgressiveStatsCollector and StatsCollector. 

- dumpStat() is identical
- startCollection() and endCollection() are very similar
- Constructors are identical

Is it possible to have ProgressiveStatsCollector extend StatsCollector, or find another way to reduce the amount of duplication?

Reply:
Heh, I knew this comment would come up.  Initially I didn't even know if we'd keep both collectors.  dumpStat() isn't identical, but it's close...I could break it up.  I'll look into maybe making ProgressiveStatsCollector subclass StatsCollector, although they're really just two different implementations of the same thing.  Maybe they should both implement the same interface, but that wouldn't help the code duplication unless I also subclass.

I guess I wasn't sure if it was going to be worth it, but I'll give it a look and see if I can make it work.  I don't like code duplication either.

Reply:
I made the necessary changes to have ProgressiveStatsCollector subclass StatsCollector.